server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java

```diff
@@ -122,13 +127,18 @@ protected RowRecord nextWithoutConstraint() throws IOException {
 
     long[] timestampArray = new long[timeStampFetchSize];
     int timeArrayLength = 0;
-    if (hasCachedTimestamp) {
+    while (!cachedTimestamps.isEmpty()) {
+      long timestamp = cachedTimestamps.remove();
       if (timestamp < curEndTime) {
+        if (!groupByTimePlan.isAscending() && timestamp < curStartTime) {
+          cachedTimestamps.addFirst(timestamp);
+          return constructRowRecord(aggregateResultList);
+        }
         if (timestamp >= curStartTime) {
-          hasCachedTimestamp = false;
           timestampArray[timeArrayLength++] = timestamp;
```

Comment:
It seems there will be IndexOutOfArrayException here.
What if `cachedTimestamps`'s size is larger than `timeStampFetchSize`. And also, you call the `constructTimeArrayForOneCal` method in the following and fetch another `timeStampFetchSize` timestamp, i think it should be changed in case of `desc`.

Reply:
fixed